{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b5ca4cf0",
   "metadata": {},
   "source": [
    "### Load the Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2e064742",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "28fc62e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Primary libraries\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "72fbad93",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Secondary libraries\n",
    "from tqdm import tqdm\n",
    "# from scipy import stats\n",
    "import scipy\n",
    "# from glob import glob\n",
    "import os\n",
    "# from matplotlib import dates as mpl_dates\n",
    "# import datetime\n",
    "# from datetime import date\n",
    "# import matplotlib.patches as mpatches\n",
    "# from matplotlib import cm\n",
    "# from colorspacious import cspace_converter\n",
    "# from collections import OrderedDict\n",
    "# import matplotlib.ticker as ticker"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1349bd2d",
   "metadata": {},
   "source": [
    "### Load the files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0173d7f",
   "metadata": {},
   "source": [
    "We are dealing with .gb files with single records each.\n",
    "\n",
    "From (https://warwick.ac.uk/fac/sci/moac/people/students/peter_cock/python/genbank/)-\n",
    "\n",
    "    Depending on the type of GenBank file(s) you are interested in, they will either contain a single record, or multiple records. You can easily determine this by looking at the raw file - each record will start with a LOCUS line, followed by various other header lines, usually a list of features, the sequence data, and ends with a // line (slash slash)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4f1af666",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converted 1 records\n"
     ]
    }
   ],
   "source": [
    "count = SeqIO.convert(\"./genbank_files/bacillus_subtilis.gb\", \"genbank\",\n",
    "                      \"./converted_fasta_files/bacillus_subtilis.fasta\", \"fasta\")\n",
    "print(f'Converted {count} records')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "d36132f8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Name NC_000964, 9074 features\n"
     ]
    }
   ],
   "source": [
    "# Loading the genbank file (given it has only one record; otherwise looping over it is required)\n",
    "\n",
    "gb_record = SeqIO.read(open(\"./genbank_files/bacillus_subtilis.gb\", 'r'), \"genbank\")\n",
    "print(f'Name {gb_record.name}, {len(gb_record.features)} features')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0195503d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "type: gene\n",
      "location: [4183444:4183897](-)\n",
      "qualifiers:\n",
      "    Key: db_xref, Value: ['GeneID:937872']\n",
      "    Key: gene, Value: ['yybA']\n",
      "    Key: locus_tag, Value: ['BSU_40710']\n",
      "    Key: old_locus_tag, Value: ['BSU40710']\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(gb_record.features[9000])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "8b85ebe6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-1\n",
      "1\n"
     ]
    }
   ],
   "source": [
    "print(gb_record.features[9000].strand)\n",
    "print(gb_record.features[0].strand)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "bbd0d77b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['source', 'gene', 'CDS', 'rRNA', 'tRNA', 'misc_RNA', 'misc_feature', 'ncRNA']\n"
     ]
    }
   ],
   "source": [
    "l = []\n",
    "\n",
    "for feature in gb_record.features:\n",
    "    if feature.type not in l:\n",
    "        l.append(feature.type)\n",
    "\n",
    "print(l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "e08c2315",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4536\n"
     ]
    }
   ],
   "source": [
    "genes = []\n",
    "\n",
    "for feature in gb_record.features:\n",
    "    if feature.type == 'gene':\n",
    "        genes.append(feature)\n",
    "\n",
    "print(len(genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "bda50210",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'list'>\n",
      "['dnaN']\n",
      "dnaN\n",
      "939970\n"
     ]
    }
   ],
   "source": [
    "print(type(genes[1].qualifiers['gene']))\n",
    "print(genes[1].qualifiers['gene'])\n",
    "print(genes[1].qualifiers['gene'][0])\n",
    "print(genes[1].qualifiers['db_xref'][0][7:])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70162a6b",
   "metadata": {},
   "source": [
    "Locations provided by **BioPython** is optimum for python purposes.\n",
    "- start and end provided by it are 1938 and 3075 respectively (0-based indexing and it assumes that end position is not included),\n",
    "- while in our actual file it is 1939 and 3075 (1-based indexing and it assumes that both start and end positions are included).\n",
    "\n",
    "This way we can directly slice seq string using locations provided to obtain the seq for the features of our interest."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "8e77dd1b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "type: gene\n",
      "location: [1938:3075](+)\n",
      "qualifiers:\n",
      "    Key: db_xref, Value: ['GeneID:939970']\n",
      "    Key: gene, Value: ['dnaN']\n",
      "    Key: locus_tag, Value: ['BSU_00020']\n",
      "    Key: old_locus_tag, Value: ['BSU00020']\n",
      "\n",
      "[1938:3075](+)\n",
      "ATGAAATTCACGATTCAAAAAGATCGTCTTGTTGAAAGTGTCCAAGATGTATTAAAAGCAGTTTCATCCAGAACCACGATTCCCATTCTGACTGGTATTAAAATTGTTGCATCAGATGATGGAGTATCCTTTACAGGGAGTGACTCAGATATTTCTATTGAATCCTTCATTCCAAAAGAAGAAGGAGATAAAGAAATCGTCACTATTGAACAGCCCGGAAGCATCGTTTTACAGGCTCGCTTTTTTAGTGAAATTGTAAAAAAATTGCCGATGGCAACTGTAGAAATTGAAGTCCAAAATCAGTATTTGACGATTATCCGTTCTGGTAAAGCTGAATTTAATCTAAACGGACTGGATGCTGATGAATATCCGCACTTGCCGCAGATTGAAGAGCATCATGCGATTCAGATCCCAACTGATTTGTTAAAAAATCTAATCAGACAAACAGTATTTGCAGTGTCCACCTCAGAAACACGCCCTATCTTGACAGGTGTAAACTGGAAAGTGGAGCAAAGTGAATTATTATGCACTGCAACGGATAGCCACCGTCTTGCATTAAGAAAGGCGAAACTTGATATTCCAGAAGACAGATCTTATAACGTCGTGATTCCGGGAAAAAGTTTAACTGAACTCAGCAAGATTTTAGATGACAACCAGGAACTTGTAGATATCGTCATCACAGAAACCCAAGTTCTGTTTAAAGCGAAAAACGTCTTGTTCTTCTCACGGCTTCTGGACGGGAATTATCCAGACACAACCAGCCTGATTCCGCAAGACAGCAAAACAGAAATCATTGTGAACACAAAAGAATTCCTTCAGGCCATTGATCGTGCATCTCTTTTAGCTAGAGAGGGACGCAACAACGTTGTAAAACTGTCCGCAAAACCGGCTGAATCCATTGAAATTTCTTCCAATTCGCCAGAAATCGGTAAAGTTGTGGAAGCAATTGTTGCGGATCAAATTGAAGGTGAGGAATTAAATATCTCTTTTAGTCCAAAATATATGCTGGATGCACTAAAGGTGCTTGAAGGAGCAGAAATACGCGTAAGCTTTACAGGCGCAATGAGACCTTTCTTAATTCGCACGCCGAATGATGAAACGATTGTACAGCTTATCCTTCCTGTCAGAACCTATTAA\n",
      "\n",
      "\n",
      "ATGAAATTCACGATTCAAAAAGATCGTCTTGTTGAAAGTGTCCAAGATGTATTAAAAGCAGTTTCATCCAGAACCACGATTCCCATTCTGACTGGTATTAAAATTGTTGCATCAGATGATGGAGTATCCTTTACAGGGAGTGACTCAGATATTTCTATTGAATCCTTCATTCCAAAAGAAGAAGGAGATAAAGAAATCGTCACTATTGAACAGCCCGGAAGCATCGTTTTACAGGCTCGCTTTTTTAGTGAAATTGTAAAAAAATTGCCGATGGCAACTGTAGAAATTGAAGTCCAAAATCAGTATTTGACGATTATCCGTTCTGGTAAAGCTGAATTTAATCTAAACGGACTGGATGCTGATGAATATCCGCACTTGCCGCAGATTGAAGAGCATCATGCGATTCAGATCCCAACTGATTTGTTAAAAAATCTAATCAGACAAACAGTATTTGCAGTGTCCACCTCAGAAACACGCCCTATCTTGACAGGTGTAAACTGGAAAGTGGAGCAAAGTGAATTATTATGCACTGCAACGGATAGCCACCGTCTTGCATTAAGAAAGGCGAAACTTGATATTCCAGAAGACAGATCTTATAACGTCGTGATTCCGGGAAAAAGTTTAACTGAACTCAGCAAGATTTTAGATGACAACCAGGAACTTGTAGATATCGTCATCACAGAAACCCAAGTTCTGTTTAAAGCGAAAAACGTCTTGTTCTTCTCACGGCTTCTGGACGGGAATTATCCAGACACAACCAGCCTGATTCCGCAAGACAGCAAAACAGAAATCATTGTGAACACAAAAGAATTCCTTCAGGCCATTGATCGTGCATCTCTTTTAGCTAGAGAGGGACGCAACAACGTTGTAAAACTGTCCGCAAAACCGGCTGAATCCATTGAAATTTCTTCCAATTCGCCAGAAATCGGTAAAGTTGTGGAAGCAATTGTTGCGGATCAAATTGAAGGTGAGGAATTAAATATCTCTTTTAGTCCAAAATATATGCTGGATGCACTAAAGGTGCTTGAAGGAGCAGAAATACGCGTAAGCTTTACAGGCGCAATGAGACCTTTCTTAATTCGCACGCCGAATGATGAAACGATTGTACAGCTTATCCTTCCTGTCAGAACCTATTAA\n",
      "\n",
      "\n",
      "TTAATAGGTTCTGACAGGAAGGATAAGCTGTACAATCGTTTCATCATTCGGCGTGCGAATTAAGAAAGGTCTCATTGCGCCTGTAAAGCTTACGCGTATTTCTGCTCCTTCAAGCACCTTTAGTGCATCCAGCATATATTTTGGACTAAAAGAGATATTTAATTCCTCACCTTCAATTTGATCCGCAACAATTGCTTCCACAACTTTACCGATTTCTGGCGAATTGGAAGAAATTTCAATGGATTCAGCCGGTTTTGCGGACAGTTTTACAACGTTGTTGCGTCCCTCTCTAGCTAAAAGAGATGCACGATCAATGGCCTGAAGGAATTCTTTTGTGTTCACAATGATTTCTGTTTTGCTGTCTTGCGGAATCAGGCTGGTTGTGTCTGGATAATTCCCGTCCAGAAGCCGTGAGAAGAACAAGACGTTTTTCGCTTTAAACAGAACTTGGGTTTCTGTGATGACGATATCTACAAGTTCCTGGTTGTCATCTAAAATCTTGCTGAGTTCAGTTAAACTTTTTCCCGGAATCACGACGTTATAAGATCTGTCTTCTGGAATATCAAGTTTCGCCTTTCTTAATGCAAGACGGTGGCTATCCGTTGCAGTGCATAATAATTCACTTTGCTCCACTTTCCAGTTTACACCTGTCAAGATAGGGCGTGTTTCTGAGGTGGACACTGCAAATACTGTTTGTCTGATTAGATTTTTTAACAAATCAGTTGGGATCTGAATCGCATGATGCTCTTCAATCTGCGGCAAGTGCGGATATTCATCAGCATCCAGTCCGTTTAGATTAAATTCAGCTTTACCAGAACGGATAATCGTCAAATACTGATTTTGGACTTCAATTTCTACAGTTGCCATCGGCAATTTTTTTACAATTTCACTAAAAAAGCGAGCCTGTAAAACGATGCTTCCGGGCTGTTCAATAGTGACGATTTCTTTATCTCCTTCTTCTTTTGGAATGAAGGATTCAATAGAAATATCTGAGTCACTCCCTGTAAAGGATACTCCATCATCTGATGCAACAATTTTAATACCAGTCAGAATGGGAATCGTGGTTCTGGATGAAACTGCTTTTAATACATCTTGGACACTTTCAACAAGACGATCTTTTTGAATCGTGAATTTCAT\n"
     ]
    }
   ],
   "source": [
    "print(gb_record.features[3])\n",
    "\n",
    "print(gb_record.features[3].location)\n",
    "\n",
    "# print(gb_record.seq[gb_record.features[3].location]) # throws error\n",
    "print(gb_record.seq[gb_record.features[3].location.nofuzzy_start:\n",
    "                    gb_record.features[3].location.nofuzzy_end])\n",
    "\n",
    "print('\\n')\n",
    "print(gb_record.seq[1938:3075])\n",
    "print('\\n')\n",
    "\n",
    "# To obtain the DNA sequence corresponding to the complement, use .reverse_complement()\n",
    "print(gb_record.seq[1938:3075].reverse_complement())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "3aaf922f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-1\n",
      "1\n",
      "<class 'int'>\n",
      "<class 'int'>\n"
     ]
    }
   ],
   "source": [
    "print(gb_record.features[9000].strand)\n",
    "print(gb_record.features[0].strand)\n",
    "print(type(gb_record.features[9000].strand))\n",
    "print(type(gb_record.features[0].strand))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c59e42d7",
   "metadata": {},
   "source": [
    "### Extracting desirable information"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b946b1d",
   "metadata": {},
   "source": [
    "Useful videos for the analysis done later-\n",
    "1. https://www.youtube.com/watch?v=LdQV3cbUwEE&list=PLe1-kjuYBZ05T9iHV_z60B9mpFt201ND5&index=8\n",
    "2. https://www.youtube.com/watch?v=HP7ThAj_f1E\n",
    "\n",
    "_Both videos are on Youtube @Bioinformatics Coach_"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a09ba63",
   "metadata": {},
   "source": [
    "KeyError resolution -\n",
    "\n",
    "    gene_name = gene.qualifiers['gene'][0]\n",
    "    gene_name = gene.qualifiers.get('gene',['unavailable'])[0]\n",
    "\n",
    "_Source_: https://bioinformatics.stackexchange.com/questions/15454/keyerror-when-getting-features-from-a-genbank-file-with-biopython-with-some-acce/15456#15456"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "55b6f387",
   "metadata": {},
   "outputs": [],
   "source": [
    "# To clear the environment variables stored earlier in the notebook\n",
    "del count, gb_record, l, genes, i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "id": "8980cccd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def genbank_file_reader(file_name):\n",
    "    \"\"\"Takes in genbank file name in the folder ./genbank_files/;\n",
    "    Outputs the dataframe for that genbank file\n",
    "    \"\"\"\n",
    "    gb_record = SeqIO.read(open(f\"./genbank_files/{file_name}\", 'r'), 'genbank')\n",
    "    print(f'Name {gb_record.name}, {len(gb_record.features)} features')\n",
    "    \n",
    "    data = []\n",
    "\n",
    "    allgenes = (\n",
    "        feature\n",
    "        for feature in gb_record.features\n",
    "        if feature.type == 'gene'\n",
    "    )\n",
    "\n",
    "    for gene in allgenes:\n",
    "\n",
    "        gene_name = gene.qualifiers.get('gene',['unavailable'])[0]\n",
    "        gene_ID = gene.qualifiers['db_xref'][0][7:]\n",
    "#         gene_ID = gene.qualifiers.get('db_xref', ['GeneID:unavailable'])[0][7:]        \n",
    "        start_pos = gene.location.nofuzzy_start\n",
    "    #     start_pos = gene.location.nofuzzy_start + 1 # if 1-based indexing required\n",
    "        end_pos = gene.location.nofuzzy_end\n",
    "        strand_sense = gene.strand\n",
    "\n",
    "        gene_seq = gene.extract(gb_record).seq\n",
    "\n",
    "    #     For +ve sense strand\n",
    "        if strand_sense == 1:\n",
    "            if len(gb_record.seq[:start_pos]) >= 200:\n",
    "                upstream_flank = gb_record.seq[start_pos-200:start_pos]\n",
    "            else:\n",
    "                upstream_flank = gb_record.seq[:start_pos]\n",
    "\n",
    "            if len(gb_record.seq[end_pos:]) >= 200:\n",
    "                downstream_flank = gb_record.seq[end_pos:end_pos+200]\n",
    "            else:\n",
    "                downstream_flank = gb_record.seq[end_pos:]\n",
    "    #     For -ve sense strand\n",
    "        elif strand_sense == -1:\n",
    "            if len(gb_record.seq[:start_pos]) >= 200:\n",
    "                downstream_flank = gb_record.seq[start_pos-200:start_pos].reverse_complement()\n",
    "            else:\n",
    "                downstream_flank = gb_record.seq[:start_pos].reverse_complement()\n",
    "\n",
    "            if len(gb_record.seq[end_pos:]) >= 200:\n",
    "                upstream_flank = gb_record.seq[end_pos:end_pos+200].reverse_complement()\n",
    "            else:\n",
    "                upstream_flank = gb_record.seq[end_pos:].reverse_complement()\n",
    "\n",
    "        data.append((gene_name, gene_ID, start_pos, end_pos, strand_sense, \n",
    "                     str(gene_seq), str(upstream_flank), str(downstream_flank)))    \n",
    "\n",
    "\n",
    "    # Creating an empty dataframe\n",
    "    df = pd.DataFrame(data, columns=['gene_name', \n",
    "                                     'gene_ID', \n",
    "                                     'start_position', \n",
    "                                     'end_position', \n",
    "                                     'strand_sense', \n",
    "                                     'gene_seq', \n",
    "                                     'upstream_flank', \n",
    "                                     'downstream_flank'])\n",
    "# #     Saving the dataframe to a .csv file\n",
    "    df.to_csv('./csv_files/'+file_name[:-3]+'.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b274739",
   "metadata": {},
   "source": [
    "How to iterate over a given directory:\n",
    "https://stackoverflow.com/questions/10377998/how-can-i-iterate-over-files-in-a-given-directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "id": "5b616499",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bacillus_subtilis.gb\n",
      "Name NC_000964, 9074 features\n",
      "streptococcus_pneumoniae.gb\n",
      "Name NZ_CP020549, 4328 features\n",
      "klebsiella_pneumoniae.gb\n",
      "Name NC_016845, 10894 features\n"
     ]
    }
   ],
   "source": [
    "# To do the flank calculations for all files\n",
    "directory = os.fsencode('./genbank_files/')\n",
    "\n",
    "for file in os.listdir(directory):\n",
    "    filename = os.fsdecode(file)\n",
    "    if filename.endswith('.gb'):\n",
    "        print(filename)\n",
    "        genbank_file_reader(filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "32790b33",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_bac = pd.read_csv('./csv_files/bacillus_subtilis.csv')\n",
    "df_strep = pd.read_csv('./csv_files/streptococcus_pneumoniae.csv')\n",
    "df_kleb = pd.read_csv('./csv_files/klebsiella_pneumoniae.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "id": "b6a528b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>strand_sense</th>\n",
       "      <th>gene_seq</th>\n",
       "      <th>upstream_flank</th>\n",
       "      <th>downstream_flank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>unavailable</td>\n",
       "      <td>11849782</td>\n",
       "      <td>381</td>\n",
       "      <td>822</td>\n",
       "      <td>-1</td>\n",
       "      <td>ATGGCAGACATTACACTTATCAGCGGCAGCACCCTGGGCAGTGCGG...</td>\n",
       "      <td>AGCAGGTACTTATCAACAAGATCCAAACAATCGATGAAATTCAGTC...</td>\n",
       "      <td>GAATTGTAAAGATCGTCCGATCTTCTGTGGATAACCATGCTTAAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>unavailable</td>\n",
       "      <td>11849790</td>\n",
       "      <td>921</td>\n",
       "      <td>1380</td>\n",
       "      <td>-1</td>\n",
       "      <td>ATGGAAAATTATCAGATCGACAATCTGGACCGCGGCATTCTTGATG...</td>\n",
       "      <td>ATTTAACGAAACTGATCTGGCGTTGTTTTGCGATATATGCGGTTTT...</td>\n",
       "      <td>TCGCTTTTTTTAATCCCACATTTTTCCACAGGTAGATCCCAGCTCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>unavailable</td>\n",
       "      <td>11849897</td>\n",
       "      <td>1531</td>\n",
       "      <td>2524</td>\n",
       "      <td>1</td>\n",
       "      <td>ATGAAAACCGCATATATCGCAAAACAACGCCAGATCAGTTTCGTTA...</td>\n",
       "      <td>GCGCATCAAGAATGCCGCGGTCCAGATTGTCGATCTGATAATTTTC...</td>\n",
       "      <td>CGATTAGCGCCGCCAGCGTCGCAGCAGGCGGCTACGCAGCCCGGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>unavailable</td>\n",
       "      <td>11849899</td>\n",
       "      <td>2527</td>\n",
       "      <td>3976</td>\n",
       "      <td>-1</td>\n",
       "      <td>ATGACGCTCGATATGCTGAATGTCATGCTGGCAGTCAGCGAAGAGG...</td>\n",
       "      <td>GCGCTGCCCGGCAGCGCAGCGCAGGAGAATATGCCGGCGGAAATTC...</td>\n",
       "      <td>TCGTTACAGCAGCGCGGAAACGCTTTCGCGAACCTGCGCCGGCCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>unavailable</td>\n",
       "      <td>11849901</td>\n",
       "      <td>3972</td>\n",
       "      <td>5472</td>\n",
       "      <td>-1</td>\n",
       "      <td>ATGGCTCAAACACATTTACTGGCTGAAAGAATTTCCCGGCTCAGCG...</td>\n",
       "      <td>TATAAAACAAATGTGTGCCTAGTCCCTTGGCCCACAAAAAGCGCAC...</td>\n",
       "      <td>CGCTCGATATGCTGAATGTCATGCTGGCAGTCAGCGAAGAGGGAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5399</th>\n",
       "      <td>unavailable</td>\n",
       "      <td>11850384</td>\n",
       "      <td>5329204</td>\n",
       "      <td>5329444</td>\n",
       "      <td>-1</td>\n",
       "      <td>ATGGAAAACCTGAATATGGATCTGCTGTACATGGCTGCCGCTGTGA...</td>\n",
       "      <td>CATTCTGATTGCTGGTCTGTTGCCGTGGTGGTCACAGTGGGTTCTG...</td>\n",
       "      <td>TAAGTTGTTAGAAACCTAAGCCACAGAGTTAATTAAAGAGGTATTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5400</th>\n",
       "      <td>unavailable</td>\n",
       "      <td>11850385</td>\n",
       "      <td>5329493</td>\n",
       "      <td>5330309</td>\n",
       "      <td>-1</td>\n",
       "      <td>ATGGCTTCAGAAAATATGACGCCGCAGGATTACATAGGACACCATC...</td>\n",
       "      <td>GCATACACCAGCAAAAGGCCGTGTGGCCTGGACCTTCGCCTTCGGT...</td>\n",
       "      <td>TTTTTTACCAACACTACTACGTTTTAACTGAAACAAACTGGAGACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5401</th>\n",
       "      <td>unavailable</td>\n",
       "      <td>11850386</td>\n",
       "      <td>5330316</td>\n",
       "      <td>5330697</td>\n",
       "      <td>-1</td>\n",
       "      <td>ATGTCAGTGTCGCTCTTGAGTCGAAACGTTGCTCGCAAGCTTCTGT...</td>\n",
       "      <td>CGCTCACAGAACGCTGAAATGCACGGAAAGTTGTCCGTTGGAAAAT...</td>\n",
       "      <td>AGGCATCATGGCTTCAGAAAATATGACGCCGCAGGATTACATAGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5402</th>\n",
       "      <td>unavailable</td>\n",
       "      <td>11850387</td>\n",
       "      <td>5331313</td>\n",
       "      <td>5331937</td>\n",
       "      <td>-1</td>\n",
       "      <td>GTGCTCAACAAACTTTCTCGTCTGCTGGATGAAGCGGGTATTTCGC...</td>\n",
       "      <td>ACGTATCTCCGGGATCACCCCGGCGGCGATCTCTATTTTGCTGGTT...</td>\n",
       "      <td>TTTTTATCAAAAAATGTGGAAAAAATGTGGCCCGCGCATGTTACAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5403</th>\n",
       "      <td>unavailable</td>\n",
       "      <td>11850388</td>\n",
       "      <td>5332052</td>\n",
       "      <td>5333942</td>\n",
       "      <td>-1</td>\n",
       "      <td>ATGTTTTATCAGGATCCTTTTGACGTCATTATCATTGGCGGGGGTC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TTGCGGCTGCCTGCCGCGCTGATTTCCCAGTTGCATAATCAGCGTT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5404 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_name   gene_ID  start_position  end_position  strand_sense  \\\n",
       "0     unavailable  11849782             381           822            -1   \n",
       "1     unavailable  11849790             921          1380            -1   \n",
       "2     unavailable  11849897            1531          2524             1   \n",
       "3     unavailable  11849899            2527          3976            -1   \n",
       "4     unavailable  11849901            3972          5472            -1   \n",
       "...           ...       ...             ...           ...           ...   \n",
       "5399  unavailable  11850384         5329204       5329444            -1   \n",
       "5400  unavailable  11850385         5329493       5330309            -1   \n",
       "5401  unavailable  11850386         5330316       5330697            -1   \n",
       "5402  unavailable  11850387         5331313       5331937            -1   \n",
       "5403  unavailable  11850388         5332052       5333942            -1   \n",
       "\n",
       "                                               gene_seq  \\\n",
       "0     ATGGCAGACATTACACTTATCAGCGGCAGCACCCTGGGCAGTGCGG...   \n",
       "1     ATGGAAAATTATCAGATCGACAATCTGGACCGCGGCATTCTTGATG...   \n",
       "2     ATGAAAACCGCATATATCGCAAAACAACGCCAGATCAGTTTCGTTA...   \n",
       "3     ATGACGCTCGATATGCTGAATGTCATGCTGGCAGTCAGCGAAGAGG...   \n",
       "4     ATGGCTCAAACACATTTACTGGCTGAAAGAATTTCCCGGCTCAGCG...   \n",
       "...                                                 ...   \n",
       "5399  ATGGAAAACCTGAATATGGATCTGCTGTACATGGCTGCCGCTGTGA...   \n",
       "5400  ATGGCTTCAGAAAATATGACGCCGCAGGATTACATAGGACACCATC...   \n",
       "5401  ATGTCAGTGTCGCTCTTGAGTCGAAACGTTGCTCGCAAGCTTCTGT...   \n",
       "5402  GTGCTCAACAAACTTTCTCGTCTGCTGGATGAAGCGGGTATTTCGC...   \n",
       "5403  ATGTTTTATCAGGATCCTTTTGACGTCATTATCATTGGCGGGGGTC...   \n",
       "\n",
       "                                         upstream_flank  \\\n",
       "0     AGCAGGTACTTATCAACAAGATCCAAACAATCGATGAAATTCAGTC...   \n",
       "1     ATTTAACGAAACTGATCTGGCGTTGTTTTGCGATATATGCGGTTTT...   \n",
       "2     GCGCATCAAGAATGCCGCGGTCCAGATTGTCGATCTGATAATTTTC...   \n",
       "3     GCGCTGCCCGGCAGCGCAGCGCAGGAGAATATGCCGGCGGAAATTC...   \n",
       "4     TATAAAACAAATGTGTGCCTAGTCCCTTGGCCCACAAAAAGCGCAC...   \n",
       "...                                                 ...   \n",
       "5399  CATTCTGATTGCTGGTCTGTTGCCGTGGTGGTCACAGTGGGTTCTG...   \n",
       "5400  GCATACACCAGCAAAAGGCCGTGTGGCCTGGACCTTCGCCTTCGGT...   \n",
       "5401  CGCTCACAGAACGCTGAAATGCACGGAAAGTTGTCCGTTGGAAAAT...   \n",
       "5402  ACGTATCTCCGGGATCACCCCGGCGGCGATCTCTATTTTGCTGGTT...   \n",
       "5403                                                NaN   \n",
       "\n",
       "                                       downstream_flank  \n",
       "0     GAATTGTAAAGATCGTCCGATCTTCTGTGGATAACCATGCTTAAAA...  \n",
       "1     TCGCTTTTTTTAATCCCACATTTTTCCACAGGTAGATCCCAGCTCG...  \n",
       "2     CGATTAGCGCCGCCAGCGTCGCAGCAGGCGGCTACGCAGCCCGGTG...  \n",
       "3     TCGTTACAGCAGCGCGGAAACGCTTTCGCGAACCTGCGCCGGCCAT...  \n",
       "4     CGCTCGATATGCTGAATGTCATGCTGGCAGTCAGCGAAGAGGGAAT...  \n",
       "...                                                 ...  \n",
       "5399  TAAGTTGTTAGAAACCTAAGCCACAGAGTTAATTAAAGAGGTATTG...  \n",
       "5400  TTTTTTACCAACACTACTACGTTTTAACTGAAACAAACTGGAGACT...  \n",
       "5401  AGGCATCATGGCTTCAGAAAATATGACGCCGCAGGATTACATAGGA...  \n",
       "5402  TTTTTATCAAAAAATGTGGAAAAAATGTGGCCCGCGCATGTTACAA...  \n",
       "5403  TTGCGGCTGCCTGCCGCGCTGATTTCCCAGTTGCATAATCAGCGTT...  \n",
       "\n",
       "[5404 rows x 8 columns]"
      ]
     },
     "execution_count": 187,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_kleb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "b8732a24",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Name NC_000964, 9074 features\n"
     ]
    }
   ],
   "source": [
    "# # Loading the genbank file (given it has only one record; otherwise looping over it is required)\n",
    "\n",
    "# gb_record = SeqIO.read(open(\"./genbank_files/bacillus_subtilis.gb\", 'r'), 'genbank')\n",
    "# print(f'Name {gb_record.name}, {len(gb_record.features)} features')\n",
    "\n",
    "# # print(f'Total genes in this file are {len(allgenes)}')\n",
    "\n",
    "# data = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "41395a88",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# allgenes = (\n",
    "#     feature\n",
    "#     for feature in gb_record.features\n",
    "#     if feature.type == 'gene'\n",
    "# )\n",
    "\n",
    "# for gene in allgenes:\n",
    "# #     gene_name = gene.qualifiers['gene'][0]\n",
    "#     gene_name = gene.qualifiers.get('gene',['unavailable'])[0]\n",
    "#     start_pos = gene.location.nofuzzy_start\n",
    "# #     start_pos = gene.location.nofuzzy_start + 1 # if 1-based indexing required\n",
    "#     end_pos = gene.location.nofuzzy_end\n",
    "#     strand_sense = gene.strand\n",
    "    \n",
    "#     gene_seq = gene.extract(gb_record).seq\n",
    "# #     print(gene_name, start_pos, end_pos, strand_sense)\n",
    "# #     print(f'\\n\\n{gene_seq}')\n",
    "    \n",
    "# #     For +ve sense strand\n",
    "#     if strand_sense == 1:\n",
    "#         if len(gb_record.seq[:start_pos]) >= 200:\n",
    "#             upstream_flank = gb_record.seq[start_pos-200:start_pos]\n",
    "#         else:\n",
    "#             upstream_flank = gb_record.seq[:start_pos]\n",
    "        \n",
    "#         if len(gb_record.seq[end_pos:]) >= 200:\n",
    "#             downstream_flank = gb_record.seq[end_pos:end_pos+200]\n",
    "#         else:\n",
    "#             downstream_flank = gb_record.seq[end_pos:]\n",
    "# #     For -ve sense strand\n",
    "#     elif strand_sense == -1:\n",
    "#         if len(gb_record.seq[:start_pos]) >= 200:\n",
    "#             downstream_flank = gb_record.seq[start_pos-200:start_pos].reverse_complement()\n",
    "#         else:\n",
    "#             downstream_flank = gb_record.seq[:start_pos].reverse_complement()\n",
    "        \n",
    "#         if len(gb_record.seq[end_pos:]) >= 200:\n",
    "#             upstream_flank = gb_record.seq[end_pos:end_pos+200].reverse_complement()\n",
    "#         else:\n",
    "#             upstream_flank = gb_record.seq[end_pos:].reverse_complement()\n",
    "    \n",
    "# #     print(f'\\n{upstream_flank}\\n\\n{downstream_flank}')\n",
    "# #     data.append((gene_name, str(start_pos), str(end_pos), \n",
    "# #                 str(strand_sense), str(gene_seq), str(upstream_flank), str(downstream_flank)))\n",
    "#     data.append((gene_name, start_pos, end_pos, strand_sense, \n",
    "#                  str(gene_seq), str(upstream_flank), str(downstream_flank)))    \n",
    "# #     break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "9370ec5c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>strand_sense</th>\n",
       "      <th>gene_seq</th>\n",
       "      <th>upstream_flank</th>\n",
       "      <th>downstream_flank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>dnaA</td>\n",
       "      <td>409</td>\n",
       "      <td>1750</td>\n",
       "      <td>1</td>\n",
       "      <td>ATGGAAAATATATTAGACCTGTGGAACCAAGCCCTTGCTCAAATCG...</td>\n",
       "      <td>ATAAGTTGTGGATTGATTTCACACAGCTTGTGTAGAAGGTTGTCCA...</td>\n",
       "      <td>CAGGACCGGGGATCAATCGGGGAAAGTGTGAATAACTTTTCGGAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>dnaN</td>\n",
       "      <td>1938</td>\n",
       "      <td>3075</td>\n",
       "      <td>1</td>\n",
       "      <td>ATGAAATTCACGATTCAAAAAGATCGTCTTGTTGAAAGTGTCCAAG...</td>\n",
       "      <td>CAGCTTAAATAGCAGGACCGGGGATCAATCGGGGAAAGTGTGAATA...</td>\n",
       "      <td>TCCGATACACTGCTGCCGACCCGTCGGCAGCTTTTCTATTCGGTAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rlbA</td>\n",
       "      <td>3205</td>\n",
       "      <td>3421</td>\n",
       "      <td>1</td>\n",
       "      <td>ATGGCAAATCCGATTTCAATTGATACAGAGATGATTACACTCGGAC...</td>\n",
       "      <td>ACCTTTCTTAATTCGCACGCCGAATGATGAAACGATTGTACAGCTT...</td>\n",
       "      <td>AGCGGGTGACACTGATTGTATATCCAGAACTTAGAACTGACATCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>recF</td>\n",
       "      <td>3436</td>\n",
       "      <td>4549</td>\n",
       "      <td>1</td>\n",
       "      <td>TTGTATATCCAGAACTTAGAACTGACATCTTACCGCAACTACGACC...</td>\n",
       "      <td>TGATTACACTCGGACAATTCTTAAAATTAGCCGATGTGATTCAGTC...</td>\n",
       "      <td>AGAAATGAGGTGAGCAATTGTATATTCATTTAGGTGATGACTTTGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>remB</td>\n",
       "      <td>4566</td>\n",
       "      <td>4812</td>\n",
       "      <td>1</td>\n",
       "      <td>TTGTATATTCATTTAGGTGATGACTTTGTGGTTTCAACACGAGATA...</td>\n",
       "      <td>CCCATTTTACTATTGGATGATGTACTGAGTGAACTGGATGATTATC...</td>\n",
       "      <td>AAATTTTTTATCACGAATATATCGTTTAGAAAAGTGTAGGTGAATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4531</th>\n",
       "      <td>mnmE</td>\n",
       "      <td>4211509</td>\n",
       "      <td>4212889</td>\n",
       "      <td>-1</td>\n",
       "      <td>ATGGATACAATTGCGGCAATTTCCACACCGATGGGAGAAGGCGCGA...</td>\n",
       "      <td>TATCCTGTTGATAAAAAGGAAGTTCAGGAAATAAAATTAATTGTTA...</td>\n",
       "      <td>TTAAAGGAGGAACTAGAATCATGGGGTATGAAGCAGGCCAATACGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4532</th>\n",
       "      <td>jag</td>\n",
       "      <td>4213199</td>\n",
       "      <td>4213826</td>\n",
       "      <td>-1</td>\n",
       "      <td>GTGAGGAATGTGACTGCTGCAGGGCGAAATGTCGATGAAGCAGTAC...</td>\n",
       "      <td>AATGCGCAGCAAAATCCGCAAATGGCGATGATGCTTTGGATTATGC...</td>\n",
       "      <td>CATAAAACCGAAGTCCGATAAAAATTGGATTTCGGTTTTTTTGTAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4533</th>\n",
       "      <td>oxaAA</td>\n",
       "      <td>4213822</td>\n",
       "      <td>4214608</td>\n",
       "      <td>-1</td>\n",
       "      <td>ATGTTGTTGAAAAGGAGAATAGGGTTGCTATTAAGTATGGTTGGCG...</td>\n",
       "      <td>GTCTGCAGCATCTATTCAGAAAGTCTTCATTATATAAGAAATCTTC...</td>\n",
       "      <td>GGAATGTGACTGCTGCAGGGCGAAATGTCGATGAAGCAGTACAGTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4534</th>\n",
       "      <td>rnpA</td>\n",
       "      <td>4214752</td>\n",
       "      <td>4215103</td>\n",
       "      <td>-1</td>\n",
       "      <td>TTGAAGAAGCGAAATCGTTTAAAGAAAAATGAAGATTTTCAAAAAG...</td>\n",
       "      <td>AGTTTTAGCACGCCGTCGCCGCAAAGGCAGAAAAGTATTATCAGCT...</td>\n",
       "      <td>GCATCACCACTGCTTTCTTACAACTTTTAAAAGCGCTTCCTTTCTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4535</th>\n",
       "      <td>rpmH</td>\n",
       "      <td>4215254</td>\n",
       "      <td>4215389</td>\n",
       "      <td>-1</td>\n",
       "      <td>ATGAAAAGAACATTCCAACCGAATAACCGTAAACGCAGCAAAGTTC...</td>\n",
       "      <td>TAGGGGCAAATTGTTTGAATTTGTCTTTTTTCCGTAAAATTAAGGA...</td>\n",
       "      <td>GCCACTGAATAATGTCAGTGGTCTTTTTTCACGTTATGGGAAAAGA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4536 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_name  start_position  end_position  strand_sense  \\\n",
       "0         dnaA             409          1750             1   \n",
       "1         dnaN            1938          3075             1   \n",
       "2         rlbA            3205          3421             1   \n",
       "3         recF            3436          4549             1   \n",
       "4         remB            4566          4812             1   \n",
       "...        ...             ...           ...           ...   \n",
       "4531      mnmE         4211509       4212889            -1   \n",
       "4532       jag         4213199       4213826            -1   \n",
       "4533     oxaAA         4213822       4214608            -1   \n",
       "4534      rnpA         4214752       4215103            -1   \n",
       "4535      rpmH         4215254       4215389            -1   \n",
       "\n",
       "                                               gene_seq  \\\n",
       "0     ATGGAAAATATATTAGACCTGTGGAACCAAGCCCTTGCTCAAATCG...   \n",
       "1     ATGAAATTCACGATTCAAAAAGATCGTCTTGTTGAAAGTGTCCAAG...   \n",
       "2     ATGGCAAATCCGATTTCAATTGATACAGAGATGATTACACTCGGAC...   \n",
       "3     TTGTATATCCAGAACTTAGAACTGACATCTTACCGCAACTACGACC...   \n",
       "4     TTGTATATTCATTTAGGTGATGACTTTGTGGTTTCAACACGAGATA...   \n",
       "...                                                 ...   \n",
       "4531  ATGGATACAATTGCGGCAATTTCCACACCGATGGGAGAAGGCGCGA...   \n",
       "4532  GTGAGGAATGTGACTGCTGCAGGGCGAAATGTCGATGAAGCAGTAC...   \n",
       "4533  ATGTTGTTGAAAAGGAGAATAGGGTTGCTATTAAGTATGGTTGGCG...   \n",
       "4534  TTGAAGAAGCGAAATCGTTTAAAGAAAAATGAAGATTTTCAAAAAG...   \n",
       "4535  ATGAAAAGAACATTCCAACCGAATAACCGTAAACGCAGCAAAGTTC...   \n",
       "\n",
       "                                         upstream_flank  \\\n",
       "0     ATAAGTTGTGGATTGATTTCACACAGCTTGTGTAGAAGGTTGTCCA...   \n",
       "1     CAGCTTAAATAGCAGGACCGGGGATCAATCGGGGAAAGTGTGAATA...   \n",
       "2     ACCTTTCTTAATTCGCACGCCGAATGATGAAACGATTGTACAGCTT...   \n",
       "3     TGATTACACTCGGACAATTCTTAAAATTAGCCGATGTGATTCAGTC...   \n",
       "4     CCCATTTTACTATTGGATGATGTACTGAGTGAACTGGATGATTATC...   \n",
       "...                                                 ...   \n",
       "4531  TATCCTGTTGATAAAAAGGAAGTTCAGGAAATAAAATTAATTGTTA...   \n",
       "4532  AATGCGCAGCAAAATCCGCAAATGGCGATGATGCTTTGGATTATGC...   \n",
       "4533  GTCTGCAGCATCTATTCAGAAAGTCTTCATTATATAAGAAATCTTC...   \n",
       "4534  AGTTTTAGCACGCCGTCGCCGCAAAGGCAGAAAAGTATTATCAGCT...   \n",
       "4535  TAGGGGCAAATTGTTTGAATTTGTCTTTTTTCCGTAAAATTAAGGA...   \n",
       "\n",
       "                                       downstream_flank  \n",
       "0     CAGGACCGGGGATCAATCGGGGAAAGTGTGAATAACTTTTCGGAAG...  \n",
       "1     TCCGATACACTGCTGCCGACCCGTCGGCAGCTTTTCTATTCGGTAT...  \n",
       "2     AGCGGGTGACACTGATTGTATATCCAGAACTTAGAACTGACATCTT...  \n",
       "3     AGAAATGAGGTGAGCAATTGTATATTCATTTAGGTGATGACTTTGT...  \n",
       "4     AAATTTTTTATCACGAATATATCGTTTAGAAAAGTGTAGGTGAATG...  \n",
       "...                                                 ...  \n",
       "4531  TTAAAGGAGGAACTAGAATCATGGGGTATGAAGCAGGCCAATACGA...  \n",
       "4532  CATAAAACCGAAGTCCGATAAAAATTGGATTTCGGTTTTTTTGTAT...  \n",
       "4533  GGAATGTGACTGCTGCAGGGCGAAATGTCGATGAAGCAGTACAGTC...  \n",
       "4534  GCATCACCACTGCTTTCTTACAACTTTTAAAAGCGCTTCCTTTCTG...  \n",
       "4535  GCCACTGAATAATGTCAGTGGTCTTTTTTCACGTTATGGGAAAAGA...  \n",
       "\n",
       "[4536 rows x 7 columns]"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# # Creating an empty dataframe\n",
    "# df = pd.DataFrame(data, columns=['gene_name',\n",
    "#                            'start_position',\n",
    "#                            'end_position',\n",
    "#                            'strand_sense', \n",
    "#                            'gene_seq', \n",
    "#                            'upstream_flank', \n",
    "#                            'downstream_flank'])\n",
    "\n",
    "# df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "d9e20968",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>strand_sense</th>\n",
       "      <th>gene_seq</th>\n",
       "      <th>upstream_flank</th>\n",
       "      <th>downstream_flank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>313</th>\n",
       "      <td>unavailable</td>\n",
       "      <td>281674</td>\n",
       "      <td>281809</td>\n",
       "      <td>-1</td>\n",
       "      <td>TTGGCAAGAAGCATTATCTTGAAACAAGAACAGACTCATAATGGAA...</td>\n",
       "      <td>TCTTTATGTATTGCCGAAATGGAATAATATGATTGCAACTTTGGCT...</td>\n",
       "      <td>GAAAAGAACAAACCGAATGCCGCTAACGATAACGGGATATACACAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>335</th>\n",
       "      <td>unavailable</td>\n",
       "      <td>300524</td>\n",
       "      <td>300656</td>\n",
       "      <td>-1</td>\n",
       "      <td>ATGACAAAGGAGGGCGGAGTTTCCAGAAAATCCGGTGAAAACCGTG...</td>\n",
       "      <td>TTTCTTCACAAAGGCGGCCCCTTTCATCCTTCTTAAGTTCCCTCCA...</td>\n",
       "      <td>ACCATTTGAGGAGGAAAAATGAATGTTTCAGAAAAAAACGTACGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>397</th>\n",
       "      <td>unavailable</td>\n",
       "      <td>369019</td>\n",
       "      <td>369217</td>\n",
       "      <td>1</td>\n",
       "      <td>TTGTCCAAGGATGAATGCCAGACGTACCAAGGAAACGCATGTGAAA...</td>\n",
       "      <td>ACTGAATATAACGGCTTTGGAATGCATGAATGATTTCACCTATAGT...</td>\n",
       "      <td>ACCATAATGGGGAGTCAAATGAATATTTACAAACCAGCCGGATTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>502</th>\n",
       "      <td>unavailable</td>\n",
       "      <td>490545</td>\n",
       "      <td>490749</td>\n",
       "      <td>-1</td>\n",
       "      <td>TTGATAAGAATAAGCATGCAAGGACAAATGCATCTTCCTTTGAAAC...</td>\n",
       "      <td>GAGTCGCGCCCGGCATTGCCATGATCGTGCTCAGTATCATTTCCCT...</td>\n",
       "      <td>CCGTTTCAAACTTTTCTTCAGTGACGGCATTTCAAATTTTTGGTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>544</th>\n",
       "      <td>unavailable</td>\n",
       "      <td>524359</td>\n",
       "      <td>524476</td>\n",
       "      <td>1</td>\n",
       "      <td>ATGTGGAATTATGAATACACAGCGATAGCTGATGTGGATGTGGGAA...</td>\n",
       "      <td>CAATCGGTAGAAGAAATATCGAATTCGCTGAAAATGTATACGACAT...</td>\n",
       "      <td>AAGGACAATTTCAAAATGGAACGGAAGGAACATTGCTTCTGCAAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4104</th>\n",
       "      <td>unavailable</td>\n",
       "      <td>3804713</td>\n",
       "      <td>3804875</td>\n",
       "      <td>-1</td>\n",
       "      <td>ATGTTTTTGTCCTCGCTTCCTTCAAAACTCATTAAGACAATCCTTT...</td>\n",
       "      <td>GGGTGTGAGCCCGCTTTAACTTTTGCAGAAAAGCGTCAACTCTATC...</td>\n",
       "      <td>GATATTGACTATGAAAAGCGTGGTGTTTTTATGAAAGACGTATCTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4155</th>\n",
       "      <td>unavailable</td>\n",
       "      <td>3856781</td>\n",
       "      <td>3856937</td>\n",
       "      <td>-1</td>\n",
       "      <td>ATGCTTTTTACGGTTTCCAGAGAGGGAAGGACAGCTGAGATCTTCC...</td>\n",
       "      <td>TTGGTATGCTGGTGCCGGGTTTGCGTTTTTGATGATGGGAGCGGCA...</td>\n",
       "      <td>TCGGGAATTTGGGTGGAACCACGGATGATCAACACATTCGTCCCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4251</th>\n",
       "      <td>unavailable</td>\n",
       "      <td>3951823</td>\n",
       "      <td>3951958</td>\n",
       "      <td>-1</td>\n",
       "      <td>ATGGAATGTTTCGAGCTGATTCATAATAAAGGAAAAGATACCTGGA...</td>\n",
       "      <td>TGTCCACTTGACAGCGGGTTTTTCATACAGCTGTTTCATTAATTGC...</td>\n",
       "      <td>TGAAATCAGGGCGAGAAGCGTATAATAAGAAAGAAAACGCCTTCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4452</th>\n",
       "      <td>unavailable</td>\n",
       "      <td>4147418</td>\n",
       "      <td>4147547</td>\n",
       "      <td>-1</td>\n",
       "      <td>TTGCCAGGCTGTTTTGATATGCAAAAAAAACTGAACAGGGGTCACG...</td>\n",
       "      <td>GCTGAAGGAACTGGATGTCATCACGGAATTTGACGGATACAAAGTA...</td>\n",
       "      <td>CTGCAAATACTTCCGTTATTTGCATGGACTTTAATGGTTTGTACCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4483</th>\n",
       "      <td>unavailable</td>\n",
       "      <td>4171800</td>\n",
       "      <td>4171986</td>\n",
       "      <td>-1</td>\n",
       "      <td>TTGTCACATCAAACAAAATACCCCGTCAATGAAGCCGCAATAATGC...</td>\n",
       "      <td>AGCTCATTTTCTTTTTATGACAGCTGAATATCTTGAAAAACAGCAC...</td>\n",
       "      <td>TATATCCAATCATGATGATTAGGGGAATTCATTTGGAACAGCAATT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_name  start_position  end_position  strand_sense  \\\n",
       "313   unavailable          281674        281809            -1   \n",
       "335   unavailable          300524        300656            -1   \n",
       "397   unavailable          369019        369217             1   \n",
       "502   unavailable          490545        490749            -1   \n",
       "544   unavailable          524359        524476             1   \n",
       "...           ...             ...           ...           ...   \n",
       "4104  unavailable         3804713       3804875            -1   \n",
       "4155  unavailable         3856781       3856937            -1   \n",
       "4251  unavailable         3951823       3951958            -1   \n",
       "4452  unavailable         4147418       4147547            -1   \n",
       "4483  unavailable         4171800       4171986            -1   \n",
       "\n",
       "                                               gene_seq  \\\n",
       "313   TTGGCAAGAAGCATTATCTTGAAACAAGAACAGACTCATAATGGAA...   \n",
       "335   ATGACAAAGGAGGGCGGAGTTTCCAGAAAATCCGGTGAAAACCGTG...   \n",
       "397   TTGTCCAAGGATGAATGCCAGACGTACCAAGGAAACGCATGTGAAA...   \n",
       "502   TTGATAAGAATAAGCATGCAAGGACAAATGCATCTTCCTTTGAAAC...   \n",
       "544   ATGTGGAATTATGAATACACAGCGATAGCTGATGTGGATGTGGGAA...   \n",
       "...                                                 ...   \n",
       "4104  ATGTTTTTGTCCTCGCTTCCTTCAAAACTCATTAAGACAATCCTTT...   \n",
       "4155  ATGCTTTTTACGGTTTCCAGAGAGGGAAGGACAGCTGAGATCTTCC...   \n",
       "4251  ATGGAATGTTTCGAGCTGATTCATAATAAAGGAAAAGATACCTGGA...   \n",
       "4452  TTGCCAGGCTGTTTTGATATGCAAAAAAAACTGAACAGGGGTCACG...   \n",
       "4483  TTGTCACATCAAACAAAATACCCCGTCAATGAAGCCGCAATAATGC...   \n",
       "\n",
       "                                         upstream_flank  \\\n",
       "313   TCTTTATGTATTGCCGAAATGGAATAATATGATTGCAACTTTGGCT...   \n",
       "335   TTTCTTCACAAAGGCGGCCCCTTTCATCCTTCTTAAGTTCCCTCCA...   \n",
       "397   ACTGAATATAACGGCTTTGGAATGCATGAATGATTTCACCTATAGT...   \n",
       "502   GAGTCGCGCCCGGCATTGCCATGATCGTGCTCAGTATCATTTCCCT...   \n",
       "544   CAATCGGTAGAAGAAATATCGAATTCGCTGAAAATGTATACGACAT...   \n",
       "...                                                 ...   \n",
       "4104  GGGTGTGAGCCCGCTTTAACTTTTGCAGAAAAGCGTCAACTCTATC...   \n",
       "4155  TTGGTATGCTGGTGCCGGGTTTGCGTTTTTGATGATGGGAGCGGCA...   \n",
       "4251  TGTCCACTTGACAGCGGGTTTTTCATACAGCTGTTTCATTAATTGC...   \n",
       "4452  GCTGAAGGAACTGGATGTCATCACGGAATTTGACGGATACAAAGTA...   \n",
       "4483  AGCTCATTTTCTTTTTATGACAGCTGAATATCTTGAAAAACAGCAC...   \n",
       "\n",
       "                                       downstream_flank  \n",
       "313   GAAAAGAACAAACCGAATGCCGCTAACGATAACGGGATATACACAA...  \n",
       "335   ACCATTTGAGGAGGAAAAATGAATGTTTCAGAAAAAAACGTACGCT...  \n",
       "397   ACCATAATGGGGAGTCAAATGAATATTTACAAACCAGCCGGATTTT...  \n",
       "502   CCGTTTCAAACTTTTCTTCAGTGACGGCATTTCAAATTTTTGGTTT...  \n",
       "544   AAGGACAATTTCAAAATGGAACGGAAGGAACATTGCTTCTGCAAGG...  \n",
       "...                                                 ...  \n",
       "4104  GATATTGACTATGAAAAGCGTGGTGTTTTTATGAAAGACGTATCTA...  \n",
       "4155  TCGGGAATTTGGGTGGAACCACGGATGATCAACACATTCGTCCCTT...  \n",
       "4251  TGAAATCAGGGCGAGAAGCGTATAATAAGAAAGAAAACGCCTTCTT...  \n",
       "4452  CTGCAAATACTTCCGTTATTTGCATGGACTTTAATGGTTTGTACCG...  \n",
       "4483  TATATCCAATCATGATGATTAGGGGAATTCATTTGGAACAGCAATT...  \n",
       "\n",
       "[67 rows x 7 columns]"
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# df[df.gene_name == 'unavailable']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "1d1245f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# del data\n",
    "# data = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11d0d9fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # 'gene_seq' can be extracted using string splicing with added .reverse_complement() for -ve sense strands\n",
    "# # Way 1:\n",
    "# if .strand == 1:\n",
    "#     gene_seq = gb_record.seq[gb_record.features[3].location.nofuzzy_start:gb_record.features[3].location.nofuzzy_end]\n",
    "# elif .strand == -1:\n",
    "#     gene_seq = gb_record.seq[gb_record.features[3].location.nofuzzy_start:gb_record.features[3].location.nofuzzy_end].reverse_complement()\n",
    "# # Way 2:\n",
    "# gene.extract(gb_record).seq\n",
    "\n",
    "# ----------------------------------------------------------------------\n",
    "\n",
    "# # For +ve sense strand\n",
    "# for +ve sense strand, upstream_flank is obtained by \n",
    "# if len(gb_record.seq[:gb_record.features[3].location.nofuzzy_start] >= 200):\n",
    "#     [gb_record.features[3].location.nofuzzy_start-200:gb_record.features[3].location.nofuzzy_start]\n",
    "# else:\n",
    "#     [:gb_record.features[3].location.nofuzzy_start]\n",
    "    \n",
    "# for +ve sense strand, downstream_flank is obtained by \n",
    "# if len(gb_record.seq[gb_record.features[3].location.nofuzzy_end:] >= 200):\n",
    "#     [gb_record.features[3].location.nofuzzy_end:gb_record.features[3].location.nofuzzy_end+200]\n",
    "# else:\n",
    "#     [gb_record.features[3].location.nofuzzy_end:]\n",
    "    \n",
    "    \n",
    "# # 'upstream_flank' and 'downstream_flank'; caution required for -ve sense strand\n",
    "# # For -ve sense strand\n",
    "# for -ve sense strand, upstream_flank is obtained by \n",
    "# if len(gb_record.seq[gb_record.features[3].location.nofuzzy_end:] >= 200):\n",
    "#     [gb_record.features[3].location.nofuzzy_end:gb_record.features[3].location.nofuzzy_end+200].reverse_complement()\n",
    "# else:\n",
    "#     [gb_record.features[3].location.nofuzzy_end:].reverse_complement()\n",
    "    \n",
    "# for -ve sense strand, downstream_flank is obtained by \n",
    "# if len(gb_record.seq[:gb_record.features[3].location.nofuzzy_start] >= 200):\n",
    "#     [gb_record.features[3].location.nofuzzy_start-200:gb_record.features[3].location.nofuzzy_start].reverse_complement()\n",
    "# else:\n",
    "#     [:gb_record.features[3].location.nofuzzy_start].reverse_complement()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "01be5892",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "yaaC 14846 15794 -1\n",
      "\n",
      "ATGACATATCATGAGTGGAAAGACTTAGCGCTATTTTACTCCGTTGAATCAACCCAGAAATTTCTAGAAAAGGTTTATATCTTAAACGGTATTAACGATGCGAAAAAAAACTCTTTTAAAAACAGTGAACGGTTTATTTATTTCTTAAAGCATGCCGAATCGTTTTATAAACAAGCCGCATATTCTCCCTTAGAAATCAAACCGATTCTTTTATTTTATGGAATGGCACAGCTTATCAAAGCATGCTTAATTACACGTGATCCTCATTATCCAAGCCATACATCAGTACTGGCACATGGCGTAACAACAAGAAAACGAAAGAAACAAAATTATTGTTTTAGCGACGATGAAGTGAAAATACAACGGAATGGCCTATGCGTACATTTTATGAAACATTTATTCGGACAATCGGATATAGTAGATGAACGGTATACCATGAAAAAATTATTAATGGCCATTCCAGAACTTAGCGATATTTTCTACTTTCAGCAGAAAGAACGATTTATGACAAAAGTTGAGAAGGATAAGAACGAAATTTTTGTACCGGAGGAAGTAGTAATCAATTACAAAATGTCTGATTCAAGGTTTGCTGAATACATGTCCCATCATTATCAATGGTCTTTTACAAAAAAAAATGAACACGGGCTGCTGTTCGAAATTTCACCACAAGACAAGGAGCCATGGACATCGACCAGCTTATTGTTTGACATGGAAAAGAACCAATATTATATCCCTTCTCAAAGGGAGCAATTTCTGAGGCTTCCCGAAATGACAATTCATTATTTGATTTTGTACAACGTTGGGATGATTGCAAGGTATGAAACAGAATGGTGGTATGAGTTATTAACACAGCATATAAGTGATGACTACGTGTTAATTCAGCAGTTTTTATTAGTATCAGAAAAGAAATTCCCTAAATACGCTTCACAGTTTCTTCTTCATTTTTAA\n",
      "\n",
      "TTAAAAATGAAGAAGAAACTGTGAAGCGTATTTAGGGAATTTCTTTTCTGATACTAATAAAAACTGCTGAATTAACACGTAGTCATCACTTATATGCTGTGTTAATAACTCATACCACCATTCTGTTTCATACCTTGCAATCATCCCAACGTTGTACAAAATCAAATAATGAATTGTCATTTCGGGAAGCCTCAGAAATTGCTCCCTTTGAGAAGGGATATAATATTGGTTCTTTTCCATGTCAAACAATAAGCTGGTCGATGTCCATGGCTCCTTGTCTTGTGGTGAAATTTCGAACAGCAGCCCGTGTTCATTTTTTTTTGTAAAAGACCATTGATAATGATGGGACATGTATTCAGCAAACCTTGAATCAGACATTTTGTAATTGATTACTACTTCCTCCGGTACAAAAATTTCGTTCTTATCCTTCTCAACTTTTGTCATAAATCGTTCTTTCTGCTGAAAGTAGAAAATATCGCTAAGTTCTGGAATGGCCATTAATAATTTTTTCATGGTATACCGTTCATCTACTATATCCGATTGTCCGAATAAATGTTTCATAAAATGTACGCATAGGCCATTCCGTTGTATTTTCACTTCATCGTCGCTAAAACAATAATTTTGTTTCTTTCGTTTTCTTGTTGTTACGCCATGTGCCAGTACTGATGTATGGCTTGGATAATGAGGATCACGTGTAATTAAGCATGCTTTGATAAGCTGTGCCATTCCATAAAATAAAAGAATCGGTTTGATTTCTAAGGGAGAATATGCGGCTTGTTTATAAAACGATTCGGCATGCTTTAAGAAATAAATAAACCGTTCACTGTTTTTAAAAGAGTTTTTTTTCGCATCGTTAATACCGTTTAAGATATAAACCTTTTCTAGAAATTTCTGGGTTGATTCAACGGAGTAAAATAGCGCTAAGTCTTTCCACTCATGATATGTCAT\n",
      "\n",
      "ATGACATATCATGAGTGGAAAGACTTAGCGCTATTTTACTCCGTTGAATCAACCCAGAAATTTCTAGAAAAGGTTTATATCTTAAACGGTATTAACGATGCGAAAAAAAACTCTTTTAAAAACAGTGAACGGTTTATTTATTTCTTAAAGCATGCCGAATCGTTTTATAAACAAGCCGCATATTCTCCCTTAGAAATCAAACCGATTCTTTTATTTTATGGAATGGCACAGCTTATCAAAGCATGCTTAATTACACGTGATCCTCATTATCCAAGCCATACATCAGTACTGGCACATGGCGTAACAACAAGAAAACGAAAGAAACAAAATTATTGTTTTAGCGACGATGAAGTGAAAATACAACGGAATGGCCTATGCGTACATTTTATGAAACATTTATTCGGACAATCGGATATAGTAGATGAACGGTATACCATGAAAAAATTATTAATGGCCATTCCAGAACTTAGCGATATTTTCTACTTTCAGCAGAAAGAACGATTTATGACAAAAGTTGAGAAGGATAAGAACGAAATTTTTGTACCGGAGGAAGTAGTAATCAATTACAAAATGTCTGATTCAAGGTTTGCTGAATACATGTCCCATCATTATCAATGGTCTTTTACAAAAAAAAATGAACACGGGCTGCTGTTCGAAATTTCACCACAAGACAAGGAGCCATGGACATCGACCAGCTTATTGTTTGACATGGAAAAGAACCAATATTATATCCCTTCTCAAAGGGAGCAATTTCTGAGGCTTCCCGAAATGACAATTCATTATTTGATTTTGTACAACGTTGGGATGATTGCAAGGTATGAAACAGAATGGTGGTATGAGTTATTAACACAGCATATAAGTGATGACTACGTGTTAATTCAGCAGTTTTTATTAGTATCAGAAAAGAAATTCCCTAAATACGCTTCACAGTTTCTTCTTCATTTTTAA\n",
      "False\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "# allgenes = (\n",
    "#     feature\n",
    "#     for feature in gb_record.features\n",
    "#     if feature.type == 'gene'\n",
    "# )\n",
    "\n",
    "# for gene in allgenes:\n",
    "#     if gene.strand == -1:\n",
    "#         gene_name = gene.qualifiers['gene'][0]\n",
    "#         start_pos = gene.location.nofuzzy_start\n",
    "#     #     start_pos = gene.location.nofuzzy_start + 1 # if 1-based indexing required\n",
    "#         end_pos = gene.location.nofuzzy_end\n",
    "#         strand_sense = gene.strand\n",
    "\n",
    "#         gene_seq = gene.extract(gb_record).seq\n",
    "#         gene_seq2 = gb_record.seq[start_pos:end_pos]\n",
    "#         gene_seq3 = gb_record.seq[start_pos:end_pos].reverse_complement()\n",
    "#         print(gene_name, start_pos, end_pos, strand_sense)\n",
    "#         print(f'\\n{gene_seq}\\n\\n{gene_seq2}\\n\\n{gene_seq3}')\n",
    "#         print(gene_seq == gene_seq2)\n",
    "#         print(gene_seq == gene_seq3)\n",
    "#         break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "85d84a65",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dnaA 409 1750 1\n",
      "\n",
      "ATGGAAAATATATTAGACCTGTGGAACCAAGCCCTTGCTCAAATCGAAAAAAAGTTGAGCAAACCGAGTTTTGAGACTTGGATGAAGTCAACCAAAGCCCACTCACTGCAAGGCGATACATTAACAATCACGGCTCCCAATGAATTTGCCAGAGACTGGCTGGAGTCCAGATACTTGCATCTGATTGCAGATACTATATATGAATTAACCGGGGAAGAATTGAGCATTAAGTTTGTCATTCCTCAAAATCAAGATGTTGAGGACTTTATGCCGAAACCGCAAGTCAAAAAAGCGGTCAAAGAAGATACATCTGATTTTCCTCAAAATATGCTCAATCCAAAATATACTTTTGATACTTTTGTCATCGGATCTGGAAACCGATTTGCACATGCTGCTTCCCTCGCAGTAGCGGAAGCGCCCGCGAAAGCTTACAACCCTTTATTTATCTATGGGGGCGTCGGCTTAGGGAAAACACACTTAATGCATGCGATCGGCCATTATGTAATAGATCATAATCCTTCTGCCAAAGTGGTTTATCTGTCTTCTGAGAAATTTACAAACGAATTCATCAACTCTATCCGAGATAATAAAGCCGTCGACTTCCGCAATCGCTATCGAAATGTTGATGTGCTTTTGATAGATGATATTCAATTTTTAGCGGGGAAAGAACAAACCCAGGAAGAATTTTTCCATACATTTAACACATTACACGAAGAAAGCAAACAAATCGTCATTTCAAGTGACCGGCCGCCAAAGGAAATTCCGACACTTGAAGACAGATTGCGCTCACGTTTTGAATGGGGACTTATTACAGATATCACACCGCCTGATCTAGAAACGAGAATTGCAATTTTAAGAAAAAAGGCCAAAGCAGAGGGCCTCGATATTCCGAACGAGGTTATGCTTTACATCGCGAATCAAATCGACAGCAATATTCGGGAACTCGAAGGAGCATTAATCAGAGTTGTCGCTTATTCATCTTTAATTAATAAAGATATTAATGCTGATCTGGCCGCTGAGGCGTTGAAAGATATTATTCCTTCCTCAAAACCGAAAGTCATTACGATAAAAGAAATTCAGAGGGTAGTAGGCCAGCAATTTAATATTAAACTCGAGGATTTCAAAGCAAAAAAACGGACAAAGTCAGTAGCTTTTCCGCGTCAAATCGCCATGTACTTATCAAGGGAAATGACTGATTCCTCTCTTCCTAAAATCGGTGAAGAGTTTGGAGGACGTGATCATACGACCGTTATTCATGCGCATGAAAAAATTTCAAAACTGCTGGCAGATGATGAACAGCTTCAGCAGCATGTAAAAGAAATTAAAGAACAGCTTAAATAG\n",
      "\n",
      "ATGGAAAATATATTAGACCTGTGGAACCAAGCCCTTGCTCAAATCGAAAAAAAGTTGAGCAAACCGAGTTTTGAGACTTGGATGAAGTCAACCAAAGCCCACTCACTGCAAGGCGATACATTAACAATCACGGCTCCCAATGAATTTGCCAGAGACTGGCTGGAGTCCAGATACTTGCATCTGATTGCAGATACTATATATGAATTAACCGGGGAAGAATTGAGCATTAAGTTTGTCATTCCTCAAAATCAAGATGTTGAGGACTTTATGCCGAAACCGCAAGTCAAAAAAGCGGTCAAAGAAGATACATCTGATTTTCCTCAAAATATGCTCAATCCAAAATATACTTTTGATACTTTTGTCATCGGATCTGGAAACCGATTTGCACATGCTGCTTCCCTCGCAGTAGCGGAAGCGCCCGCGAAAGCTTACAACCCTTTATTTATCTATGGGGGCGTCGGCTTAGGGAAAACACACTTAATGCATGCGATCGGCCATTATGTAATAGATCATAATCCTTCTGCCAAAGTGGTTTATCTGTCTTCTGAGAAATTTACAAACGAATTCATCAACTCTATCCGAGATAATAAAGCCGTCGACTTCCGCAATCGCTATCGAAATGTTGATGTGCTTTTGATAGATGATATTCAATTTTTAGCGGGGAAAGAACAAACCCAGGAAGAATTTTTCCATACATTTAACACATTACACGAAGAAAGCAAACAAATCGTCATTTCAAGTGACCGGCCGCCAAAGGAAATTCCGACACTTGAAGACAGATTGCGCTCACGTTTTGAATGGGGACTTATTACAGATATCACACCGCCTGATCTAGAAACGAGAATTGCAATTTTAAGAAAAAAGGCCAAAGCAGAGGGCCTCGATATTCCGAACGAGGTTATGCTTTACATCGCGAATCAAATCGACAGCAATATTCGGGAACTCGAAGGAGCATTAATCAGAGTTGTCGCTTATTCATCTTTAATTAATAAAGATATTAATGCTGATCTGGCCGCTGAGGCGTTGAAAGATATTATTCCTTCCTCAAAACCGAAAGTCATTACGATAAAAGAAATTCAGAGGGTAGTAGGCCAGCAATTTAATATTAAACTCGAGGATTTCAAAGCAAAAAAACGGACAAAGTCAGTAGCTTTTCCGCGTCAAATCGCCATGTACTTATCAAGGGAAATGACTGATTCCTCTCTTCCTAAAATCGGTGAAGAGTTTGGAGGACGTGATCATACGACCGTTATTCATGCGCATGAAAAAATTTCAAAACTGCTGGCAGATGATGAACAGCTTCAGCAGCATGTAAAAGAAATTAAAGAACAGCTTAAATAG\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "# allgenes = (\n",
    "#     feature\n",
    "#     for feature in gb_record.features\n",
    "#     if feature.type == 'gene'\n",
    "# )\n",
    "\n",
    "# for gene in allgenes:\n",
    "#     gene_name = gene.qualifiers['gene'][0]\n",
    "#     start_pos = gene.location.nofuzzy_start\n",
    "# #     start_pos = gene.location.nofuzzy_start + 1 # if 1-based indexing required\n",
    "#     end_pos = gene.location.nofuzzy_end\n",
    "#     strand_sense = gene.strand\n",
    "    \n",
    "#     gene_seq = gene.extract(gb_record).seq\n",
    "#     gene_seq2 = gb_record.seq[start_pos:end_pos]\n",
    "#     print(gene_name, start_pos, end_pos, strand_sense)\n",
    "#     print(f'\\n{gene_seq}\\n\\n{gene_seq2}')\n",
    "#     print(gene_seq == gene_seq2)\n",
    "#     break"
   ]
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
